tests/cli/test_cli.py

```diff
@@ -0,0 +1,265 @@
+    bob_public_keys = MockSideChannel.BobPublicKeys(bob_view_response['result']['bob_encrypting_key'],
+                                                    bob_view_response['result']['bob_verifying_key'])
+
+    side_channel.save_bob_public_keys(bob_public_keys)
+
+    """
+    Scene 3: Alice creates a policy, and saves the policy metadata to a sidechannel.
+    """
+
+    random_label = random_policy_label.decode()  # Unicode string
+
+    derive_args = ('--mock-networking',
+                   '--json-ipc',
+                   'alice', 'derive-policy',
+                   '--config-file', alice_configuration_file_location,
+                   '--label', random_label)
+
+    derive_response = click_runner.invoke(nucypher_cli, derive_args, catch_exceptions=False, env=envvars)
+    assert derive_response.exit_code == 0
+
+    derive_response = json.loads(derive_response.output)
+    assert derive_response['result']['label'] == random_label
+
+    # Alice and the sidechannel: at Tinagre
+    policy = MockSideChannel.PolicyAndLabel(encrypting_key=derive_response['result']['policy_encrypting_key'],
+                                            label=derive_response['result']['label'])
+    side_channel.save_policy(policy=policy)
+
+    """
+    Scene 4: We catch up with Alice later on, after she has learned about existing Ursulas
+    """
+    teacher = list(federated_ursulas)[0]
+    teacher_uri = teacher.seed_node_metadata(as_teacher_uri=True)
+
+    # Some Ursula is running somewhere
+    def _run_teacher():
+        start_pytest_ursula_services(ursula=teacher)
+        return teacher_uri
+
+    def _grant(teacher_uri):
+
+        # Alice fetched Bob's public keys from the side channel
+        bob_keys = side_channel.fetch_bob_public_keys()
+        bob_encrypting_key = bob_keys.bob_encrypting_key
+        bob_verifying_key = bob_keys.bob_verifying_key
+
+        grant_args = ('--mock-networking',
+                      '--json-ipc',
+                      'alice', 'grant',
+                      '--federated-only',
+                      '--teacher-uri', teacher_uri,
+                      '--config-file', alice_configuration_file_location,
+                      '--m', 2,
+                      '--n', 3,
+                      '--label', random_label,
+                      '--bob-encrypting-key', bob_encrypting_key,
+                      '--bob-verifying-key', bob_verifying_key)
+
+        grant_result = click_runner.invoke(nucypher_cli, grant_args, catch_exceptions=False, env=envvars)
+        assert grant_result.exit_code == 0
+
+        grant_result = json.loads(grant_result.output)
+
+        # # Alice puts the Treasure Map somewhere Bob can get it.
+        # side_channel.save_treasure_map(treasure_map=grant_result['result']['treasure_map'])
+
+        return grant_result
+
+    def enrico_encrypts(grant_result):
+        """
+        Scene 5: Enrico encrpyts some data for the policy and saves it to a side channel
+        """
```

Comment:
I'd change the order of Scene 4 (granting) and 5 (encrypting) to emphasize that encrypting is independent of granting. Also `encrpyts` --> `encrypts`

Reply:
:heavy_check_mark: - Yeah this really helped communicate the idea, Thanks.